### PR TITLE
New events client to support Streaming structured events.

### PR DESCRIPTION
**Pull request by renukamanavalan**
Refer [HLD #954](https://github.com/sonic-net/SONiC/pull/954)

### CODE REVIEW

File: sonic_data_client/events_client.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    wg          *sync.WaitGroup // wait for all sub go routines to finish
+
+    subs_handle unsafe.Pointer
+
+    stopped     int
+
+    // Stats counter
+    counters    map[string]uint64
+
+    last_latencies  [LATENCT_LIST_SIZE]int64;
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:61]: Not matching spelling on line 30

File: sonic_data_client/events_client.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+    for evtc.stopped == 0 {
+        var tmp_counters map[string]uint64
+
+        // compute latency
+        if evtc.last_latency_full {
+            var total uint64 = 0
+            var cnt int = 0
+
+            for _ v := range evtc.last_latencies {
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:133]: missing comma. for _, v

Comment (zbud-msft) [reply sonic_data_client/events_client.go:133]: Avoid one letter var names

File: sonic_data_client/events_client.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                if v > 0 {
+                    total += v
+                    cnt++
+                }
+            }
+            evtc.counters[LATENCY] = (int64) (total/cnt/1000/1000)
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:139]: What happens if last_latencies are all 0, wouldnt this be divide by 0 error?

File: sonic_data_client/events_client.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        DialTimeout:0,
+    })
+
+    var db_counters map[string]uint64
+    var wr_counters *map[string]uint64 = NULL;
+
+    /* Init current values for cumulative keys and clear for absolute */
+    for _, key := range STATS_CUMULATIVE_KEYS {
+        fv, err := rclient.HGetAll(key).Result()
+        db_counters[key] = fv[STATS_FIELD_NAME]
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:119]: Deal with error case

File: sonic_data_client/events_client.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    rclient := redis.NewClient(&redis.Options{
+        Network:    "tcp",
+        Addr:       sdcfg.GetDbTcpAddr("COUNTERS_DB", ns),
+        Password:   "", // no password set
+        DB:         sdcfg.GetDbId("COUNTERS_DB", ns)
+        DialTimeout:0,
+    })
+
+    var db_counters map[string]uint64
+    var wr_counters *map[string]uint64 = NULL;
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:114]: No semicolon

File: sonic_data_client/events_client.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+    /* Init current values for cumulative keys and clear for absolute */
+    for _, key := range STATS_CUMULATIVE_KEYS {
+        fv, err := rclient.HGetAll(key).Result()
+        db_counters[key] = fv[STATS_FIELD_NAME]
+    }
+    for _, key := range STATS_ABSOLUTE_KEYS {
+        db_counters[key] = 0;
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:122]: No semicolon

File: sonic_data_client/events_client.go (hunk 5 above)
Comment (zbud-msft) [reply sonic_data_client/events_client.go:114]: uninitialized value is nil by default

File: sonic_data_client/events_client.go [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            }
+            evtc.counters[LATENCY] = (int64) (total/cnt/1000/1000)
+        }
+
+        for key, val := range evtc.counters {
+            tmp_counters[key] = val + db_counters[key]
+        }
+        tmp_counters[DROPPED] += evtc.last_errors
+
+        if (wr_counters != nil) && !reflect.DeepEqual(tmp_counters, *wr_counters) {
```
Comment (zbud-msft) [inline sonic_data_client/events_client.go:147]: Can you add some comments for this block of code? When is wr_counters not nil? Is line 157 supposed to be outside of this block?

File: sonic_data_client/events_client.go (hunk 1 above)
Comment (renukamanavalan) [reply sonic_data_client/events_client.go:61]: The update is not reflecting. Let me check

**Pull request closed without merging**